rust/tw_internet_computer/src/protocol/request_id.rs

```diff
@@ -0,0 +1,295 @@
+// Copyright © 2017-2023 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+use std::collections::BTreeMap;
+
+use serde::{Deserialize, Serialize};
+use tw_hash::{sha2::sha256, H256};
+
+use super::envelope::{EnvelopeContent, Label};
+
+const DOMAIN_IC_REQUEST: &[u8; 11] = b"\x0Aic-request";
+
+/// When signing requests or querying the status of a request
+/// (see Request status) in the state tree, the user identifies
+/// the request using a request id, which is the
+/// representation-independent hash of the content map of the
+/// original request. A request id must have length of 32 bytes.
+pub struct RequestId(pub(crate) H256);
+
+impl RequestId {
+    /// Create the prehash from the request ID.
+    /// See: https://internetcomputer.org/docs/current/references/ic-interface-spec#envelope-authentication
+    pub fn sig_data(&self) -> H256 {
+        let mut sig_data = vec![];
+        sig_data.extend_from_slice(DOMAIN_IC_REQUEST);
+        sig_data.extend_from_slice(self.0.as_slice());
+        H256::try_from(sha256(&sig_data).as_slice()).unwrap_or_else(|_| H256::new())
+    }
+}
+
+impl From<&EnvelopeContent> for RequestId {
+    fn from(value: &EnvelopeContent) -> Self {
+        match value {
+            EnvelopeContent::Call {
+                nonce,
+                ingress_expiry,
+                sender,
+                canister_id,
+                method_name,
+                arg,
+            } => representation_independent_hash_call(
+                canister_id.as_slice().to_vec(),
+                method_name,
+                arg.clone(),
+                *ingress_expiry,
+                sender.as_slice().to_vec(),
+                nonce.as_deref(),
+            ),
+            EnvelopeContent::ReadState {
+                ingress_expiry,
+                sender,
+                paths,
+            } => representation_independent_hash_read_state(
+                *ingress_expiry,
+                paths,
+                sender.as_slice().to_vec(),
+                None,
+            ),
```

Comment:
What do you think about adding `EnvelopeCall` and `EnvelopeReadState` structures, add make these `representation_independent_hash_call`, `representation_independent_hash_read_state` methods of the structs correspondingly?